Classify C++ structs as loadable or address-only

MForster opened the pull request:
C++ structs are only loadable if they are trivially copyable.

Resolves SR-12472.

gribozavr commented:
@swift-ci Please review

gribozavr commented:
@swift-ci Please test

CodaFi commented:
This needs explicit initialization in `StructDecl::StructDecl`

MForster commented:
> This needs explicit initialization in `StructDecl::StructDecl`

Done.

martinboehme commented on include/swift/AST/Decl.h:
```diff
@@ -558,10 +558,12 @@ class alignas(1 << DeclAlignInBits) Decl {
-    HasUnreferenceableStorage : 1
+    HasUnreferenceableStorage : 1,
+    /// True if this struct is imported from C++ and not trivially copyable.
+    HasNonTrivialCopyOrMove : 1
```
Removing the negation seems like it would simplify the semantics? I.e. `TriviallyCopyable`, or even `CxxTriviallyCopyable`. Using the "trivially copyable" wording (which is a C++ term) would also emphasize that this is exactly the C++ concept.

MForster replied:
I considered that for exactly the reasons that you mention. I went with the current semantics anyway, because it is not _only_ the C++ concept. It is "imported *AND* not trivially copyable". If I invert the semantics it becomes "not imported or trivially copyable", which is a mouthful and still not a nice concept. I could go with "trivially copyable" alone, but then what does that mean for the non-C++ case? Or with "imported and trivially copyable", but then I need to check "imported" when I use the bit...

Also the concepts combine in a nicer way: if either of those bits is set it isn't loadable any more...

martinboehme replied:
> It is "imported _AND_ not trivially copyable".

Hm... good point. I have to admit though that the "is imported" bit wasn't really clear to me when I wrote the comment above... does it make sense to have this be part of the name?

Also, maybe we can still keep the C++ term "trivially copyable" and say "not trivially copyable"?

In sum, this would give us maybe `CxxNotTriviallyCopyable` or `ImportedNotTriviallyCopyable`?

MForster replied:
I changed the name to `IsCxxNotTriviallyCopyable`.

martinboehme replied:
SG

MForster commented:
@swift-ci Please test

gribozavr commented:
@swift-ci Please test macOS

gribozavr commented:
@swift-ci Please test Windows

gribozavr commented:
@swift-ci Please test Windows

gribozavr commented on test/Interop/Cxx/class/loadable-types-silgen.swift:
```diff
@@ -0,0 +1,93 @@
+}
+
+// CHECK-LABEL: sil hidden [ossa] @$s4main32passStructWithDefaultConstructor{{[_0-9a-zA-Z]*}}F
+func passStructWithDefaultConstructor(s: StructWithDefaultConstructor) {
```
You could reduce duplication in function names by calling all functions just `pass`. You will still find the type of the argument in the mangled name, so it would be possible to find the `CHECK-LABEL` line that uniquely identifies a function.

MForster replied:
What you said was not quite true, because some structs mangle differently. For example `EmptyStruct` get's mangled into something that doesn't contain that name. I simplified it in a different way.

MForster replied:
Correction: I fell victim to Swift's compression in mangling. I'd still keep it like it is.

gribozavr commented on test/Interop/Cxx/class/loadable-types-silgen.swift:
```diff
@@ -0,0 +1,93 @@
+// RUN: %target-swift-emit-silgen -I %S/Inputs -enable-cxx-interop %s | %FileCheck %s
+
```
Please add a short description of the intent of this test. Something like "This test checks that we classify C++ types as loadable and address-only correctly."

MForster replied:
Done

gribozavr commented on test/Interop/Cxx/class/Inputs/loadable-types.h:
```diff
@@ -0,0 +1,65 @@
+};
+
+struct StructWithAdditionalConstructor {
+  StructWithAdditionalConstructor(){};
+  StructWithAdditionalConstructor(int parameter){};
```
```suggestion
  StructWithAdditionalConstructor() {}
  StructWithAdditionalConstructor(int parameter) {}
```

MForster replied:
This is what `clang-format --style=LLVM` does, though...

gribozavr replied:
Ah OK then -- just drop the semicolon then.

gribozavr commented on test/Interop/Cxx/class/Inputs/loadable-types.h:
```diff
@@ -0,0 +1,65 @@
+};
+
+struct StructWithDestructor {
+  ~StructWithDestructor(){};
```
```suggestion
  ~StructWithDestructor() {}
```

gribozavr commented on test/Interop/Cxx/class/Inputs/loadable-types.h:
```diff
@@ -0,0 +1,65 @@
+
+struct StructWithContainedDestructor {
+  StructWithDestructor member;
+};
```
Please add these tests:

```c++
struct StructTriviallyCopyableMovable {
  StructTriviallyCopyableMovable(const StructTriviallyCopyableMovable&) = default;
  StructTriviallyCopyableMovable(StructTriviallyCopyableMovable&&) = default;
  StructTriviallyCopyableMovable &operator=(const StructTriviallyCopyableMovable&) = default;
  StructTriviallyCopyableMovable &operator=(StructTriviallyCopyableMovable&&) = default;
  ~StructTriviallyCopyableMovable() = default;
};

struct StructNonCopyableTriviallyMovable {
  StructNonCopyableTriviallyMovable(const StructNonCopyableTriviallyMovable&) = delete;
  StructNonCopyableTriviallyMovable(StructNonCopyableTriviallyMovable&&) = default;
  StructNonCopyableTriviallyMovable &operator=(const StructNonCopyableTriviallyMovable&) = delete;
  StructNonCopyableTriviallyMovable &operator=(StructNonCopyableTriviallyMovable&&) = default;
  ~StructNonCopyableTriviallyMovable() = default;
};

struct StructNonCopyableNonMovable {
  StructNonCopyableNonMovable(const StructNonCopyableNonMovable&) = delete;
  StructNonCopyableNonMovable(StructNonCopyableNonMovable&&) = default;
  StructNonCopyableNonMovable &operator=(const StructNonCopyableNonMovable&) = delete;
  StructNonCopyableNonMovable &operator=(StructNonCopyableNonMovable&&) = default;
  ~StructNonCopyableNonMovable() = default;
};

struct StructDeletedDestructor {
  StructTriviallyCopyableMovable(const StructTriviallyCopyableMovable&) = default;
  StructTriviallyCopyableMovable(StructTriviallyCopyableMovable&&) = default;
  StructTriviallyCopyableMovable &operator=(const StructTriviallyCopyableMovable&) = default;
  StructTriviallyCopyableMovable &operator=(StructTriviallyCopyableMovable&&) = default;
  ~StructTriviallyCopyableMovable() = delete;
};
```

MForster replied:
Done.

gribozavr commented on test/Interop/Cxx/class/loadable-types-silgen.swift:
```diff
@@ -0,0 +1,93 @@
+import LoadableTypes
+
+// CHECK-LABEL: sil hidden [ossa] @$s4main21passAddressOnlyStruct{{[_0-9a-zA-Z]*}}F
+func passAddressOnlyStruct(s: EmptyStruct) {
```
The function name does not match the intent :)

MForster replied:
Done

gribozavr commented on test/Interop/Cxx/class/Inputs/loadable-types.h:
```diff
@@ -0,0 +1,98 @@
+
+struct StructWithInheritedMoveAssignment : StructWithMoveAssignment {};
+
+struct StructWithContainedMoveAssignment {
```
Sorry, this is a really pedantic nitpick, but could you replace "Contained" with "Subobject" throughout the patch? "subobject" is a term that is used in the C++ spec.

MForster replied:
Done.

gribozavr commented:
@swift-ci Please test

gribozavr commented:
@swift-ci Please test

gribozavr commented:
@swift-ci Please test

gribozavr commented on test/Interop/Cxx/class/loadable-types-silgen.swift:
```diff
@@ -0,0 +1,116 @@
+  // CHECK: bb0(%0 : $StructTriviallyCopyableMovable):
+}
+
+// CHECK-LABEL: sil hidden [ossa] @$s4main4pass{{.*[ (]}}StructNonCopyableTriviallyMovable)
+func pass(s: StructNonCopyableTriviallyMovable) {
+  // CHECK: bb0(%0 : $StructNonCopyableTriviallyMovable):
+}
+
+// CHECK-LABEL: sil hidden [ossa] @$s4main4pass{{.*[ (]}}StructNonCopyableNonMovable)
+func pass(s: StructNonCopyableNonMovable) {
+  // CHECK: bb0(%0 : $StructNonCopyableNonMovable):
+}
+
+// CHECK-LABEL: sil hidden [ossa] @$s4main4pass{{.*[ (]}}StructDeletedDestructor)
+func pass(s: StructDeletedDestructor) {
+  // CHECK: bb0(%0 : $StructDeletedDestructor):
+}
```
Shouldn't the last three cases be address-only?

MForster replied:
Yes. I just wasn't done, yet :-)

gribozavr commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -3468,6 +3468,11 @@ namespace {
 
+      if (auto cxxRecordDecl = dyn_cast<clang::CXXRecordDecl>(decl)) {
+        result->setIsCxxNotTriviallyCopyable(
+            !cxxRecordDecl->isTriviallyCopyable());
```
Also has trivial copy constructor?

MForster replied:
Done.

gribozavr commented:
@swift-ci Please test

shahmishal commented:
@swift-ci Please test

MForster committed:
Classify C++ structs as loadable or address-only

C++ structs are only loadable if they are trivially copyable.

Resolves SR-12472.

MForster commented:
@swift-ci Please test

gribozavr commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -3468,6 +3468,24 @@ namespace {
+
+        for (const auto *ctor : cxxRecordDecl->ctors()) {
+          if (ctor->isCopyConstructor() && ctor->isDeleted()) {
+            result->setIsCxxNotTriviallyCopyable(true);
```
Do you think we should also check that the copy constructor and destructor are public?

MForster replied:
Done

MForster commented:
@swift-ci Please test

gribozavr commented on test/Interop/Cxx/class/Inputs/loadable-types.h:
```diff
@@ -0,0 +1,158 @@
+  StructWithPrivateDefaultedDestructor subobject;
+};
+
+// Tests for common sets of special subobjects
```
```suggestion
// Tests for common sets of special member functions.
```

gribozavr commented on test/Interop/Cxx/class/Inputs/loadable-types.h:
```diff
@@ -0,0 +1,158 @@
+};
+
+struct StructWithDestructor {
+  ~StructWithDestructor(){};
```
```suggestion
  ~StructWithDestructor(){}
```

Extra semicolon.

gribozavr commented on test/Interop/Cxx/class/loadable-types-silgen.swift:
```diff
@@ -0,0 +1,180 @@
+  // CHECK: bb0(%0 : $*StructWithSubobjectPrivateDefaultedDestructor):
+}
+
+// Tests for common sets of special subobjects
```
```suggestion
// Tests for common sets of special member functions.
```

MForster committed:
Update test/Interop/Cxx/class/Inputs/loadable-types.h

Co-authored-by: Dmitri Gribenko <gribozavr@gmail.com>

MForster committed:
Update test/Interop/Cxx/class/Inputs/loadable-types.h

Co-authored-by: Dmitri Gribenko <gribozavr@gmail.com>

MForster committed:
Update test/Interop/Cxx/class/loadable-types-silgen.swift

Co-authored-by: Dmitri Gribenko <gribozavr@gmail.com>

MForster commented:
@swift-ci Please test
